contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,189 @@
+
+React's declarative approach simplifies this process because it abstracts the complexity of interacting with the actual DOM elements and events. For example, it provides its own [event system](https://reactjs.org/docs/events.html) so that events can work in the same way across different browsers.
+
+#### Benefit 2: Better Performance
+
+Web applications can have a lot of user interaction and data updates, which results in changes being made to the DOM. Adding and removing DOM nodes isn't slow, but the performance-bottleneck arises because the browser needs to *[reflow and repaint](https://www.phpied.com/rendering-repaint-reflowrelayout-restyle/)* the UI every time the DOM is changed.
+
+React minimizes this update time by using a *virtual DOM*. The virtual DOM is a JavaScript object that is kept in the memory of your application.
+
+![React's Virtual DOM](javascript-framework-react-imgs/ReactVirtualDOM.png "React's Virtual DOM")
+_Figure 1. How React's actual DOM gets updated._
+
+As shown in Figure 2 above, updates to the UI will first be made to the virtual DOM. Then, React will compare the virtual DOM with the actual DOM using a *[diffing algorithm](https://reactjs.org/docs/reconciliation.html#the-diffing-algorithm)*. 
+Finally, React updates the actual DOM only in places it differs with the virtual DOM. It batches multiple changes together and updates the actual DOM in one go, minimizing update time. 
+
+#### Benefit 3: Testability
+
+A React application is made up of a combination of components. Components are independent from each other, and like functions, they map the same input to the same output. This makes it easy to write unit tests for your application.
+
+Additionally, React only allows data to flow downwards (one-way data binding) using `state` and `props`, which makes your application [easier to debug](https://reactjs.org/docs/design-principles.html#debugging) as you can be sure that the data updates the UI, and never the other way around. 
+
+#### Other Advantages of React
+
+Besides the three main benefits explained above, React has the following advantages:
+ - React is not an <tooltip content="An opiniated framework guides or locks you into their preferred way of doing things.">*opinionated*</tooltip> framework, which gives you the flexibility to choose your application stack.
+ - As React is an open-source library backed by Facebook, it is constantly being updated with new features and performance optimizations.
```

Comment:
> As React is an open-source library backed by Facebook, it is constantly being updated with new features and performance optimizations.

It's not clear why the former results in the latter.

Reply:
Noted, I'll update this